[FrameworkBundle] [DX] Add `Controller::json` method to make it easy to send json

mcfedr opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

Its currently a awkward to use Serializer component to send a `JsonResponse`.

I have tried two approaches
1. use `Serializer::normalize` and `JsonResponse`
2. use `Serializer::serialize` and a plain `Response`, and set the `content-type`

In either cases there is need for a custom `json` function so as not to repeat yourself and there are disadvantages.
1. In the first case you are only partly using `Serializer` and any custom `Encoder` would be skipped
2. In the second you are not making use of `JsonResponse`, particular disadvantage if you want to support JSONP. 

This new `json` method uses the serializer component is enabled it is used to generate the json data, and falls back to normal `JsonResponse` when its not.

stof commented on src/Symfony/Bundle/FrameworkBundle/Response/JsonSerializedResponse.php:
```diff
@@ -0,0 +1,76 @@
+     */
+    public function setContext(array $context)
+    {
+        $this->context = $context;
```
this is broken in case you call the setter after setting data

stof commented on src/Symfony/Bundle/FrameworkBundle/Response/JsonSerializedResponse.php:
```diff
@@ -0,0 +1,76 @@
+     */
+    public function setSerializer(SerializerInterface $serializer)
+    {
+        $this->serializer = $serializer;
```
Same issue here. Btw, injecting a serializer into the class looks weird.

dunglas commented:
Why not something like https://github.com/dunglas/DunglasApiBundle/blob/1.x/Controller/ResourceController.php#L79-L93?

I agree with @stof, `Response` should be a value object. The serializer should not be injected in it. An helper method in the controller like this one API Platform would be better IMO. Btw, we should wait for a decision about #16863 before adding new methods to the controller IMO.

mcfedr commented:
Now I look at it again, I completely agree on the weirdness of inserting the Serializer into the Response.

How about what I have done now? Its very similar to how DunglasApiBundle does it, but keeping the `JsonResponse`.

dunglas commented on src/Symfony/Bundle/FrameworkBundle/Controller/Controller.php:
```diff
@@ -98,6 +100,28 @@ protected function redirectToRoute($route, array $parameters = array(), $status
+     *
+     * @return JsonResponse
+     */
+    protected function json($data, $status = 200, $headers = array(), $context = array())
```
What about `renderJson` or something similar to existing methods?

WouterJ replied:
`json` is consistent with `stream` to return a stream response.

mcfedr replied:
I was looking at `redirect` and `stream` as similar names, also `renderView` for example returns a `string`, not a `Response`

dunglas replied:
ok :)

dunglas commented on src/Symfony/Bundle/FrameworkBundle/Controller/Controller.php:
```diff
@@ -98,6 +100,28 @@ protected function redirectToRoute($route, array $parameters = array(), $status
+     */
+    protected function json($data, $status = 200, $headers = array(), $context = array())
+    {
+        if ($this->container->has('serializer')) {
```
I'm not a fan of having 2 different behaviors depending of the configuration. What do you think about splitting this method in 2 separates methods and throw an error in the method using the serializer if it is not available?

mcfedr replied:
I do understand this point, but then again, looking at `render`/`renderView`, changes according to whether you have templating or twig enabled.

Also if #17603 is merged the result would be the same after you enable the serializer.

mcfedr commented:
@dunglas That was a long read! Would certainly wait until a decision is made there. I see some real value in breaking up an every growing Controller class.

mcfedr committed:
Add a Controller function to make it easy to return json

If the serializer component is enabled it is used to generate the json
data, if not the standard `json_encode`  function is used

fabpot commented:
LGTM. ping @symfony/deciders

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Controller/Controller.php:
```diff
@@ -98,6 +99,29 @@ protected function redirectToRoute($route, array $parameters = array(), $status
+                'json_encode_options' => JsonResponse::DEFAULT_ENCODING_OPTIONS,
+            ), $context));
+
+            return new JsonResponse($json, $status, $headers, true);
```
The fourth argument should be removed.

xabbuh replied:
But you now have to increase the required version of `symfony/http-foundation`.

mcfedr commented on src/Symfony/Bundle/FrameworkBundle/composer.json:
```diff
@@ -22,7 +22,7 @@
         "symfony/config": "~2.8|~3.0",
         "symfony/event-dispatcher": "~2.8|~3.0",
-        "symfony/http-foundation": "~2.8|~3.0",
+        "symfony/http-foundation": "~3.1",
```
@xabbuh is this the best way to do it? I see at least one other component has a `~3.0` requirement, I assume this is expected to change as new functionality is added to Symfony 3

xabbuh replied:
yes, that's right

mcfedr commented on src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
+3.1.0
+-----
+
+ * Added `Controller:json` to simplify creating JSON responses when using the serializer component
```
There is such a mix of capitalisation and full stops in this fill, I hope this is the expected format.

xabbuh replied:
Serializer component

mcfedr commented:
@xabbuh Does anything else need to be done?

Taluu commented on src/Symfony/Component/HttpFoundation/JsonResponse.php:
```diff
@@ -29,16 +29,19 @@ class JsonResponse extends Response
-    protected $encodingOptions = 15;
+    const DEFAULT_ENCODING_OPTIONS = 15;
+
+    protected $encodingOptions = self::DEFAULT_ENCODING_OPTIONS;
```
IRC, the possibility to assign a property from a constant value is available only on php >= 5.6, and Sf3 is php >= 5.5...

mcfedr replied:
As far as I can tell this usage has always been part of PHP 5. It certainly works in 5.5. You made me doubt it and I ran the tests for this class in 5.5 to double check!

I know that from PHP 7 its possible to write this how I would have really liked

    const DEFAULT_ENCODING_OPTIONS = JSON_HEX_TAG | JSON_HEX_APOS | JSON_HEX_AMP | JSON_HEX_QUOT

Taluu replied:
Ow, yeah I may have mistaken with arithmetics such as `const DEFAULT_ENCODING_OPTIONS = JSON_HEX_TAG | JSON_HEX_APOS | JSON_HEX_AMP | JSON_HEX_QUOT` indeed. My bad !

xabbuh commented:
:+1:

mcfedr commented:
@xabbuh does the 'needs work' label still need to be here?

xabbuh commented:
@mcfedr No, you're right.

Status: Reviewed

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
+3.1.0
+-----
+
+ * Added `Controller:json` to simplify creating JSON responses when using the Serializer component
```
typo: missing double colon

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTest.php:
```diff
@@ -124,6 +126,85 @@ private function getContainerWithTokenStorage($token = null)
+        $serializer
+            ->expects($this->once())
+            ->method('serialize')
+            ->with(array(), 'json', array('json_encode_options' => 15))
```
you could use the const here instead of 15

mcfedr replied:
I would tend to prefer not to use the const my tests for the very reason that it will mean that if its changed it breaks the tests, and that requires fixing

fabpot replied:
Quite the contrary if you ask me. We should test that using the constant (whatever its value) does not break the test. +1 for using the constant.

mcfedr replied:
@fabpot which side are you on? I cannot tell if your comment was written before or after mine

fabpot replied:
I'm in favor of using the constant.

fabpot commented:
Thank you @mcfedr.

taylorotwell commented:
This breaks BC by adding a new property to public `setData` method of JsonResponse, which Laravel overrides.

taylorotwell commented on src/Symfony/Component/HttpFoundation/JsonResponse.php:
```diff
@@ -88,34 +91,37 @@ public function setCallback($callback = null)
      * @throws \InvalidArgumentException
      */
-    public function setData($data = array())
+    public function setData($data = array(), $preEncoded = false)
```
Breaks BC.

weaverryan commented:
I agree - it looks like this violates our BC policy - http://symfony.com/doc/current/contributing/code/bc.html#changing-classes - sub-item "Public Methods -> Add argument with a default value". I don't think we allow a change like this.

dunglas commented:
What we can do instead is adding a new public setPreEncoded method.

nicolas-grekas commented:
Or "setJson" :-) which should be called by setData internally btw

fabpot commented:
I think the `setJson()` method is indeed a good solution. @nicolas-grekas Can you work on a PR?

HeahDude commented on src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
+3.1.0
+-----
+
+ * Added `Controller::json` to simplify creating JSON responses when using the Serializer component
```
Missing `[BC break]` ?

dunglas replied:
AFAIK, it's not a BC break.

HeahDude replied:
Agreed, sorry I got confused by #18373.

avtehnik commented:
Awsome!

Etheriq commented:
nice job, Fred!
